### ISSUE TITLE
summer intern feedback and open issues

### ISSUE THREAD

**Issue by Solosneros** [acerbilab/gpyreg#8]
I just copied his notes from a file on slack to discuss them here (similar to https://github.com/lacerbi/pyvbmc/issues/28).

### Things to investigate:
 The diagnostics computed in slice sampling differ slightly from what is in pymc, what is the cause of this?
     - original MATLAB split slice sample chain into 4 parts for whatever reason, in gpyreg and pymc the split is to 2
     - pymc and gpyreg/original MATLAB have slightly different formula
  
### Improvements:
- Docstrings of slice sampling might need editing for terminology, e.g. use of "stochastic parameters"
     - this was not originally written by me, but terminology from pymc
- Posterior object is slgithly unnecessary looking

### Issues:
 - Test for bayesian quadrature with noise fails sometimes after latest bugfix, look into that a bit
 - Calling predict without giving hyperparameters crashes the library, implement an exception for this
 - Calling quad also crashes without hyperparameters, and no data GP as well
 - We can call update with too large (or too small I think) hyperparameter array, causing silent bugs
 
### OpenBLAS:
- OpenBLAS needs to have threading set to 1 for it to function well.
- export OPENBLAS_NUM_THREADS=1 fixes the problem though it restricts parallelism a bit
- this can be done by the user, or within the program as well I believe
- another way to solve this is to use another BLAS library such as Intel MKL, which is the default for Anaconda
- however, Intel MKL has another issue, namely than on machines with an AMD processor performance tanks
- this can currently be fixed on AMD machines by export MKL_DEBUG_CPU_TYPE=5
- however, there is no guarantee that this fix will work since it is based on undocumented features that are subject to change

**Comment by lacerbi** [comment]
Follow-up (*almost exactly five years later*) after checking `main` at a2f8ddce867f502e29717959cf0ff3529f598618 (after #43). Several concerns still apply. The recommendations below are based on source inspection and targeted reproductions; no fixes for these findings have been made yet.

1. **Fix noisy quadrature variance and replace the inconsistent test reference.**

   There is a reproducible correctness bug beyond the historical report of a flaky test. [`quad`](https://github.com/acerbilab/gpyreg/blob/a2f8ddce867f502e29717959cf0ff3529f598618/gpyreg/gaussian_process.py#L2066) reconstructs the Cholesky normalization from the constant-noise hyperparameter alone, while the factorization uses the minimum total training noise. User-provided noise can therefore make the integral variance incorrect, even driving it to the numerical floor.

   Reproduction: a one-dimensional squared-exponential GP with zero mean, unit length/output scales, constant noise variance `0.01`, user-provided variance `1`, `X=[[-1], [0], [1]]`, and `y=[[1], [0], [1]]`. `quad(0, 1, compute_var=True)` returns variance approximately `2.22e-16`; independent Gauss-Hermite integration of `predict_full(add_noise=False)` returns `0.200693762307239`. The means agree (`0.379164673330362`). References with 40, 60, and 80 nodes agree, and the constant-noise-only control passes.

   Recommendation: use the actual posterior factorization's normalization instead of assuming that the first noise hyperparameter describes all training noise.

   The [existing noisy quadrature test](https://github.com/acerbilab/gpyreg/blob/a2f8ddce867f502e29717959cf0ff3529f598618/gpyreg/testing/test_gaussian_process.py#L588) also compares different quantities:

   - Numerical integration uses hyperparameter sample zero; `quad` averages all ten samples.
   - The mean reference uses Gaussian standard deviation `0.11`; `quad` uses `0.1`.
   - The covariance reference includes observation noise; `quad` computes uncertainty in the latent-function integral.
   - The grid spacing uses `30/1000` although the endpoint-inclusive grid has spacing `30/999`.

   Replace this with deterministic references using fixed hyperparameters and matched sample, measure, and noise semantics. Keep stochastic fitting accuracy as a separate test. The original intermittent failure itself has not been reproduced in this investigation; the implementation defect and test inconsistencies have.

2. **Validate `update` inputs before changing GP state.**

   [`update`](https://github.com/acerbilab/gpyreg/blob/a2f8ddce867f502e29717959cf0ff3529f598618/gpyreg/gaussian_process.py#L728) still accepts extra hyperparameter columns silently. With too few columns, it can raise only after appending observations or replacing valid posterior state. Without training data, invalid widths can be stored without an error. The documented one-dimensional hyperparameter input also reaches an unpacking error.

   `set_hyperparameters` already checks column count, but direct `update` bypasses that validation. Normalize and validate the hyperparameter array at the public boundary, before any state mutation, and test that rejected updates leave the GP unchanged.

   There is a related issue in the [rank-one shortcut](https://github.com/acerbilab/gpyreg/blob/a2f8ddce867f502e29717959cf0ff3529f598618/gpyreg/gaussian_process.py#L775): appending one observation silently ignores supplied replacement hyperparameters, including valid replacements. Use the shortcut only when the existing hyperparameters remain applicable; otherwise recompute the posterior.

3. **Add explicit missing-state errors while preserving prior prediction.**

   Fresh `predict` raises `AttributeError: 'NoneType' object has no attribute 'size'`; fresh `quad` raises the corresponding `shape` error. These should become informative exceptions explaining the required initialization. Supplying data through default `update` without hyperparameters can also leave NaN hyperparameters and produce NaN outputs.

   Valid hyperparameters with no observations already support prior prediction, which should remain supported. For prior-only quadrature, explicitly choose between implementing the prior integral and rejecting that state clearly.

4. **Fix invalid sampler diagnostics before deciding whether to modernize them.**

   Differences from historical PyMC are not, by themselves, defects. GPyReg's R-hat is the classical split estimator; historical PyMC included an additional finite-sample variance correction. Modern ArviZ uses rank-normalized/folded split R-hat, so numerical agreement requires matching the estimator and chain arrangement. See the [ArviZ reference](https://python.arviz.org/en/v0.22.0/api/generated/arviz.rhat.html) and [Stan's discussion](https://mc-stan.org/docs/2_38/reference-manual/analysis.html).

   Two current defects were reproduced:

   - The [ESS truncation](https://github.com/acerbilab/gpyreg/blob/a2f8ddce867f502e29717959cf0ff3529f598618/gpyreg/slice_sample.py#L851) can return negative effective sample sizes: four ordinary draws give ESS `-4`; `np.tile([-1., 1.], 10)[:, None]` gives ESS `-20`. Termination on a negative autocorrelation pair and termination by exhausting the trace need different handling.
   - Constant traces produce `R=NaN` and `ESS=NaN`, yet [`__diagnose`](https://github.com/acerbilab/gpyreg/blob/a2f8ddce867f502e29717959cf0ff3529f598618/gpyreg/slice_sample.py#L618) reports success because comparisons with NaN do not trigger its failure conditions. Handle insufficient samples, undefined diagnostics, and known fixed coordinates explicitly.

   Add deterministic short, constant, anticorrelated, and positively correlated fixtures. Treat migration to rank-normalized R-hat and bulk/tail ESS as a separate design decision, with comparisons against a pinned reference implementation. ESS greater than the number of draws can be valid for anticorrelated samples; negative ESS cannot.

   These diagnostics are enabled by default for standalone `SliceSampler`; `GP.fit` explicitly disables them.

   Regarding MATLAB: the bundled wrapper passes two halves to an external `psrf` function. That helper is absent from this repository, so the claim of four-way splitting remains unverified; it could result from that helper splitting each input again.

5. **Clarify terminology; retain the purpose of `Posterior`.**

   Replace wording such as “dimension of the stochastic” with “number of sampled parameters.” “Sampled parameter” is clearer here than “stochastic parameter.”

   `Posterior` has a useful role: it stores coefficients and matrix factors for one hyperparameter setting, allowing repeated predictions without recomputing them. Better documentation is preferable to removing this container solely because it is small.

6. **Update the BLAS advice to be workload- and environment-dependent.**

   One BLAS thread can help small matrices and avoid oversubscription when the application already parallelizes work, but it is not universally optimal. A small local OpenBLAS benchmark of `GP.log_likelihood` found one thread slightly faster at 115 training points (`0.141` versus `0.152` ms with four threads), while four threads were faster at 800 points (`12.4` versus `13.8` ms). These are illustrative measurements on one environment, not general performance guarantees.

   Document user-controlled thread settings and benchmark representative workloads. Avoid silently imposing a process-wide thread policy inside the package. See [OpenBLAS guidance](https://github.com/OpenMathLib/OpenBLAS/blob/develop/USAGE.md), its [runtime variables](https://www.openmathlib.org/OpenBLAS/docs/runtime_variables/), and [threadpoolctl's runtime controls and limitations](https://github.com/joblib/threadpoolctl).

   Retire the undocumented `MKL_DEBUG_CPU_TYPE=5` recommendation. Backend choice and current AMD/MKL performance should be evaluated for the actual library version, CPU, and workload rather than relying on that historical workaround.

*(Suggestions by GPT 6 Astra while I was doing other work for #43; parked here for later.)*

**Comment by lacerbi** [comment]
Status after #48 and #49 (branch `fix-quad-heteroskedastic-variance`):

**Addressed in #49**
- Item 1: `quad` variance with user-provided noise, with deterministic Gauss-Hermite reference tests replacing the inconsistent noisy quadrature test. Also found and fixed in review: `quad` read `SquaredExponentialIsotropic` hyperparameters with the ARD layout, so it was wrong for `D > 1`.
- Item 2, second part: the rank-one `update` shortcut now requires `hyp=None`, so replacement hyperparameters passed with a single new point are applied.
- Item 4, the two reproduced defects: ESS is positive and bounded (Geyer initial positive sequence with the Stan/ArviZ floor); NaN/undefined diagnostics and fixed coordinates no longer pass as success; fewer than eight recorded samples return `exit_flag -3`.
- Items 5 and 6: sampled-parameter terminology, `Posterior` attributes documented, BLAS threading section in the docs.

**Still open from the list above**
- Item 2, first part: validate the hyperparameter array in `update` before any state mutation (extra columns are still accepted silently; too few columns raise late).
- Item 3: informative errors for `predict` and `quad` on an uninitialized GP; decide on prior-only quadrature.
- Item 4, design decision: rank-normalized R-hat and bulk/tail ESS against a pinned reference implementation.

**Pre-existing issues surfaced during review of #49, not changed there**
- `update`: appending one point without `s2_new` to a GP that has `s2` leaves `self.s2` one row shorter than `X`, and the rank-one extension normalizes the new row differently from the existing factor. Candidate fix: take the shortcut only when `self.s2 is None`, or append the noise row.
- `SliceSampler.sample`: the `exit_flag == 0` "no violations detected" message is unreachable, since the diagnostics return 1 on success; `eff_N` is first bound to the iteration count and later rebound to the diagnostics array.
- `SliceSampler.__init__`: list-typed `LB`/`UB` make `LB != UB` a scalar, so fixed coordinates are not detected (the same applies to the existing `widths[LB == UB]`).

Posted via Claude Code.
